src/Parsing/Impl/Ast/ScopeInfo.cs

```diff
@@ -0,0 +1,297 @@
+        /// Gets the variables for this scope.
+        /// </summary>
+        public ICollection<PythonVariable> ScopeVariables => Variables?.Values ?? Array.Empty<PythonVariable>() as ICollection<PythonVariable>;
+
+        internal virtual bool IsGlobal => false;
+
+        public PythonAst GlobalParent {
+            get {
+                var cur = Node;
+                while (!(cur is PythonAst)) {
+                    Debug.Assert(cur != null);
+                    cur = cur.ParentNode;
+                }
+
+                return (PythonAst) cur;
+            }
+        }
+
+        internal void Clear() {
+            _references?.Clear();
+            _cellVars?.Clear();
+            _freeVars?.Clear();
+            _globalVars?.Clear();
+            _nonLocalVars?.Clear();
+        }
+
+        internal void AddFreeVariable(PythonVariable variable, bool accessedInScope) {
+            _freeVars = _freeVars ?? new List<PythonVariable>();
+            if (!_freeVars.Contains(variable)) {
+                _freeVars.Add(variable);
+            }
+        }
+
+        internal string AddReferencedGlobal(string name) {
+            _globalVars = _globalVars ?? new List<string>();
+            if (!_globalVars.Contains(name)) {
+                _globalVars.Add(name);
+            }
+
+            return name;
+        }
+
+        internal void AddNonLocalVariable(NameExpression name) {
+            _nonLocalVars = _nonLocalVars ?? new List<NameExpression>();
+            _nonLocalVars.Add(name);
+        }
+
+        internal void AddCellVariable(PythonVariable variable) {
+            _cellVars = _cellVars ?? new List<string>();
+            if (!_cellVars.Contains(variable.Name)) {
+                _cellVars.Add(variable.Name);
+            }
+        }
+
+        /// <summary>
+        /// Variables that are bound in an outer scope - but not a global scope
+        /// </summary>
+        public IReadOnlyList<PythonVariable> FreeVariables => _freeVars;
+
+        protected virtual bool ExposesLocalVariable { get; }
+
+        internal bool TryGetVariable(string name, out PythonVariable variable) {
+            if (Variables != null && name != null) {
+                return Variables.TryGetValue(name, out variable);
+            } else {
+                variable = null;
+                return false;
+            }
+        }
+
+        internal virtual bool TryBindOuter(IScopeNode from, string name, bool allowGlobals, out PythonVariable variable) {
+            // Hide scope contents by default (only functions expose their locals)
+            variable = null;
+            return false;
+        }
+
+        internal abstract PythonVariable BindReference(PythonNameBinder binder, string name);
+
+        internal virtual void Bind(PythonNameBinder binder) {
+            if (_references != null) {
+                foreach (var refList in _references.Values) {
+                    foreach (var reference in refList) {
```

Comment:
` foreach(var reference in _references.MaybeEnumerate().Values.SelectMany(v => v)`.

Reply:
ok I gave it a shot